<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Gtexv6prarevariation by joed3</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='https://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Gtexv6prarevariation</h1>
      <h2 class="project-tagline">Repository to reproduce analyses from the GTEx V6P Rare Variation Manuscript</h2>
      <a href="https://github.com/joed3/GTExV6PRareVariation" class="btn">View on GitHub</a>
      <a href="https://github.com/joed3/GTExV6PRareVariation/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/joed3/GTExV6PRareVariation/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="gtexv6prarevariation" class="anchor" href="#gtexv6prarevariation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GTExV6PRareVariation</h1>

<p>Repository to reproduce analyses from the GTEx V6P Rare Variation Manuscript</p>

<h1>
<a id="to-run-the-code" class="anchor" href="#to-run-the-code" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>To run the code</h1>

<h2>
<a id="download-required-files" class="anchor" href="#download-required-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Download required files</h2>

<p>Download from &lt;website - coming soon&gt;: <br></p>

<ul>
<li>processed data directory (referred to below as &lt;processed_data&gt;)</li>
</ul>

<p>Download from <a href="http://www.gtexportal.org/home/datasets">http://www.gtexportal.org/home/datasets</a>: <br></p>

<ul>
<li><code>gencode.v19.genes.v6p_model.patched_contigs.gtf.gz</code></li>
<li>
<code>GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_reads.gct.gz</code> (gunzip it)</li>
<li>
<code>GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_rpkm.gct.gz</code> (gunzip it)</li>
<li><code>GTEx_Data_V6_Annotations_SampleAttributesDS.txt</code></li>
<li><code>GTEx_Analysis_2015-01-12_Annotations_SubjectPhenotypesDS.txt</code></li>
<li>the covariates used during eQTL discovery</li>
</ul>

<p>Download from Gencode (<a href="http://www.gencodegenes.org/releases/19.html">http://www.gencodegenes.org/releases/19.html</a>; Comprehensive gene annotation gtf):</p>

<ul>
<li>
<code>gencode.v19.annotation.gtf.gz</code> (gunzip it)</li>
</ul>

<p>Download from dbGaP: <br>
(It is possible some of the file names may be different in the final release.)</p>

<ul>
<li><code>GTEx_Analysis_2015-01-12_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz</code></li>
<li>ASE files</li>
</ul>

<p>Download from <a href="http://krishna.gs.washington.edu/download/CADD/v1.2/">http://krishna.gs.washington.edu/download/CADD/v1.2/</a>:</p>

<ul>
<li><code>whole_genome_SNVs.tsv.gz</code></li>
<li><code>whole_genome_SNVs.tsv.gz.tbi</code></li>
<li><code>whole_genome_SNVs_inclAnno.tsv.gz</code></li>
<li><code>whole_genome_SNVs_inclAnno.tsv.gz.tbi</code></li>
</ul>

<p>Download from ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20130502/:</p>

<ul>
<li><code>ALL.chr*.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz</code></li>
</ul>

<p>Download from <a href="http://compbio.mit.edu/encode-motifs/">http://compbio.mit.edu/encode-motifs/</a>:</p>

<ul>
<li><code>matches.txt.gz</code></li>
</ul>

<p>Download Epigenomics Roadmap files from <a href="http://www.broadinstitute.org/%7Emeuleman/reg2map/HoneyBadger2_release/DNase/p2/">http://www.broadinstitute.org/~meuleman/reg2map/HoneyBadger2_release/DNase/p2/</a>:</p>

<ul>
<li><code>prom/BED_files_per_sample/regions_prom_E*.bed</code></li>
<li><code>enh/BED_files_per_sample/regions_enh_E*.bed</code></li>
<li><code>dyadic/BED_files_per_sample/regions_dyadic_E*.bed</code></li>
</ul>

<p>Download ExAC constraint data from ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3/functional_gene_constraint/:</p>

<ul>
<li><code>forweb_cleaned_exac_r03_march16_z_data_pLI.txt</code></li>
</ul>

<p>Other files: <br></p>

<ul>
<li>
<code>gtex.lumpy.gs.svscore.low_conf.vcf.gz</code> (Obtained directly from the Hall Lab.)</li>
</ul>

<h2>
<a id="setup" class="anchor" href="#setup" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Setup</h2>

<p>The code relies on an assumed directory structure.
Everything is under an upper level directory.
If you are using the processed files available online, set this upper-level directory to be the path to &lt;processed_data&gt;. <br>
Otherwise, set this to be any path where you want scripts to output.</p>

<p>Either run or add the following to your .bashrc (omit trailing slashes for all directories): <br></p>

<pre><code>export RAREVARDIR=&lt;the path to the upper-level directory&gt;
export CADD_DIR=&lt;the path to the CADD files&gt;
export KG_DIR=&lt;the path to the 1000 genomes files&gt;
export ENCODE_MOTIF_DIR=&lt;the path to matches.txt.gz (excluding the file name)&gt;
export ER_DIR=&lt;the path to the Epigenomics Roadmap directory (sub directories are prom, enh, and dyadic)&gt;
export EXAC_DIR=&lt;the path to forweb_cleaned_exac_r03_march16_z_data_pLI.txt (excluding the file name)&gt;
</code></pre>

<p>If you are going to run everything from scratch, first make the directories that are assumed to exist. <br>
You can skip this step if you will use the processed data files downloaded from &lt;website - coming soon&gt;. <br></p>

<pre><code>mkdir ${RAREVARDIR}/logs
mkdir ${RAREVARDIR}/preprocessing
mkdir ${RAREVARDIR}/preprocessing/PEER
mkdir ${RAREVARDIR}/reference
mkdir ${RAREVARDIR}/data
mkdir ${RAREVARDIR}/data/medz
mkdir ${RAREVARDIR}/data/singlez
mkdir ${RAREVARDIR}/data/metasoft
</code></pre>

<p>These directories are required to hold the results of scripts for which we do not provide processed data.</p>

<pre><code>mkdir ${RAREVARDIR}/paper_figures
mkdir ${RAREVARDIR}/features
mkdir ${RAREVARDIR}/features/variantBeds
mkdir ${RAREVARDIR}/features/annotations
mkdir ${RAREVARDIR}/features/annotations/ACMG
mkdir ${RAREVARDIR}/features/annotations/ClinVar
mkdir ${RAREVARDIR}/features/annotations/GWAS
mkdir ${RAREVARDIR}/features/annotations/OMIM
mkdir ${RAREVARDIR}/features/annotations/Orphanet
mkdir ${RAREVARDIR}/features/annotations/Other
</code></pre>

<p>Then you can run the code below. <br>
Some of the steps reproduce processed files available on &lt;website - coming soon&gt; and are marked as such.</p>

<h1>
<a id="pipeline" class="anchor" href="#pipeline" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Pipeline</h1>

<h2>
<a id="expression-data-correction-and-normalization" class="anchor" href="#expression-data-correction-and-normalization" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Expression data correction and normalization</h2>

<p>Generates processed data that can be downloaded from &lt;website - coming soon&gt;. <br></p>

<h4>
<a id="generate-rpkm-and-read-count-matrices-from-gtex-v6p-combined-file" class="anchor" href="#generate-rpkm-and-read-count-matrices-from-gtex-v6p-combined-file" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Generate rpkm and read count matrices from GTEx V6P combined file</h4>

<pre><code>cat \&lt;path to downloaded file\&gt;/GTEx_Data_V6_Annotations_SampleAttributesDS.txt | \
    cut -f1,14 | sed 's/ - /_/' | sed 's/ /_/g' | sed 's/(//' | sed 's/)//' | sed 's/c-1/c1/' | grep -v NA12878 &gt; \
    ${RAREVARDIR}/preprocessing/gtex_2015-01-12_samples_tissues.txt

python preprocessing/split_by_tissues.py \
    --GTEX \&lt;path to downloaded file\&gt;/GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_rpkm.gct \
    --SAMPLE ${RAREVARDIR}/preprocessing/gtex_2015-01-12_samples_tissues.txt \
    --OUT ${RAREVARDIR}/preprocessing/PEER \
    --END .rpkm.txt &lt;br&gt;

python preprocessing/split_by_tissues.py \
    --GTEX \&lt;path to downloaded file\&gt;/GTEx_Analysis_v6p_RNA-seq_RNA-SeQCv1.1.8_gene_reads.gct \
    --SAMPLE ${RAREVARDIR}/preprocessing/gtex_2015-1-12_samples_tissues.txt \
    --OUT ${RAREVARDIR}/preprocessing/PEER \
    --END .reads.txt
</code></pre>

<h4>
<a id="run-bash-scripts-to-generate-peer-corrected-data-includes-non-eas-with-covariates-removed" class="anchor" href="#run-bash-scripts-to-generate-peer-corrected-data-includes-non-eas-with-covariates-removed" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run bash scripts to generate PEER corrected data (includes non-EAs) with covariates removed</h4>

<p>(Uses multiple cores. Currently set to 10 cores. Can be altered by changing the number of cores <br> 
specified by <code>parallel --jobs 10</code> in the scripts <code>preprocessing/PEER/calc.PEER.factors.all.tissues.sh</code> and <br>
<code>preprocessing/PEER/calc.residuals.sh</code>) <br></p>

<pre><code>bash preprocessing/PEER/PEER.pipeline.sh
</code></pre>

<h4>
<a id="make-list-of-individuals-and-tissues" class="anchor" href="#make-list-of-individuals-and-tissues" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Make list of individuals and tissues</h4>

<pre><code>bash preprocessing/get_tissue_by_individual.sh
</code></pre>

<h4>
<a id="make-flat-files-from-raw-rpkms-and-peer-corrected-data-for-all-tissues-and-individuals" class="anchor" href="#make-flat-files-from-raw-rpkms-and-peer-corrected-data-for-all-tissues-and-individuals" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Make flat files from raw RPKMs and PEER-corrected data for all tissues and individuals</h4>

<pre><code>python preprocessing/gather_filter_normalized_expression.py
python preprocessing/gather_filter_rpkm.py
</code></pre>

<h4>
<a id="make-list-of-expressed-genes" class="anchor" href="#make-list-of-expressed-genes" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Make list of expressed genes</h4>

<pre><code>cat preprocessing/PEER/*peer.ztrans.txt | cut -f1 | sort | uniq | \
grep -v Id &gt; preprocessing/gtex.expressed.genes.txt
</code></pre>

<h4>
<a id="make-summary-statistics-for-expressed-genes" class="anchor" href="#make-summary-statistics-for-expressed-genes" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Make summary statistics for expressed genes</h4>

<p>(Uses multiple cores. Can set the number at the top of the script. Currently set to 10 cores.)</p>

<pre><code>Rscript preprocessing/rpkm.expression.analysis.R
</code></pre>

<h2>
<a id="preparing-reference-files-used-later" class="anchor" href="#preparing-reference-files-used-later" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Preparing reference files used later</h2>

<p>Generates processed data that can be downloaded from &lt;website - coming soon&gt;. <br></p>

<p>Copy or move the GTEx annotation file (<code>gencode.v19.genes.v6p_model.patched_contigs.gtf.gz</code>) to <code>${RAREVARDIR}/reference</code>.</p>

<pre><code>bash preprocessing/process.reference.files.sh \&lt;path to\&gt;/gencode.v19.genes.v6p_model.patched_contigs.gtf.gz \&lt;path to\&gt;/gencode.v19.annotation.gtf
</code></pre>

<p>(relies on <code>pad.gtf.exons.py</code>, <code>gtf2TSS.sh</code>, and <code>gtf2genebed.sh</code>)</p>

<h2>
<a id="outlier-calling" class="anchor" href="#outlier-calling" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Outlier calling</h2>

<p>Generates processed data that can be downloaded from &lt;website - coming soon&gt;. <br></p>

<h4>
<a id="call-multi-tissue-outliers" class="anchor" href="#call-multi-tissue-outliers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Call multi-tissue outliers</h4>

<p>(Uses multiple cores. Can set the number at the top of the script.)</p>

<pre><code>Rscript call_outliers/call_outliers_medz.R
</code></pre>

<h4>
<a id="call-single-tissue-outliers" class="anchor" href="#call-single-tissue-outliers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Call single-tissue outliers</h4>

<pre><code>python call_outliers/call_outliers_single_tissue.py
</code></pre>

<h4>
<a id="compare-single-tissue-and-multi-tissue-outliers-as-well-as-get-stats-on-each" class="anchor" href="#compare-single-tissue-and-multi-tissue-outliers-as-well-as-get-stats-on-each" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Compare single-tissue and multi-tissue outliers as well as get stats on each</h4>

<pre><code>Rscript call_outliers/compare_single_multi_outliers.R
</code></pre>

<h4>
<a id="run-replication-for-single-tissue-and-multi-tissue-outliers" class="anchor" href="#run-replication-for-single-tissue-and-multi-tissue-outliers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Run replication for single-tissue and multi-tissue outliers</h4>

<p>(The multi-tissue replication uses multiple cores.)</p>

<pre><code>Rscript call_outliers/multi_tissue_replication.R
Rscript call_outliers/single_tissue_replication.R
</code></pre>

<h2>
<a id="feature-generation" class="anchor" href="#feature-generation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Feature generation</h2>

<h4>
<a id="processing-vcfs-into-bed-files-for-each-individual" class="anchor" href="#processing-vcfs-into-bed-files-for-each-individual" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Processing VCFs into bed files for each individual</h4>

<pre><code>bash feature_construction/vcf2bedfiles.sh 
     \&lt;path to \&gt;/GTEx_Analysis_2015-01-12_WholeGenomeSeq_148Indiv_GATK_HaplotypeCaller.vcf.gz 
     \&lt;path to\&gt;/gtex.lumpy.gs.svscore.low_conf.vcf.gz
</code></pre>

<p>(relies on :</p>

<ul>
<li><code>vcf2bedfiles_helper_processVCF.sh</code></li>
<li><code>vcf2bedfiles_helper_processVCF_SV.sh</code></li>
<li><code>vcf2bedfiles_helper_processVCFtoolsOutput.sh</code></li>
<li><code>vcf2bedfiles_helper_processVCFtoolsOutput_CNV.sh</code></li>
<li><code>compileCADDscores.sh</code></li>
<li>
<code>extractCADDscores_ekt.py</code>)</li>
</ul>

<h4>
<a id="extract-features-to-be-combined-with-the-individual-bed-files" class="anchor" href="#extract-features-to-be-combined-with-the-individual-bed-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Extract features to be combined with the individual bed files</h4>

<pre><code>bash feature_construction/extract.1kg.AF.sh
</code></pre>

<p>(uses multiple cores; relies on <code>process.1kg.AF.py</code>)</p>

<pre><code>bash feature_construction/subset.CADD.features.sh

bash feature_construction/TFBS_pipeline.sh
</code></pre>

<p>(relies on <code>pouya.raw.summary.py</code>)</p>

<pre><code>bash feature_construction/ER_pipeline.sh
</code></pre>

<h4>
<a id="add-extracted-features-to-individual-bed-files" class="anchor" href="#add-extracted-features-to-individual-bed-files" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Add extracted features to individual bed files</h4>

<pre><code>bash run_add_features_variant_beds.sh
</code></pre>

<p><strong>Important:</strong> Make sure to use bedtools version 2.26.0 or later.
Memory leak in previous versions causes the memory for this script to blow up. <br>
(uses multiple cores; relies on <code>add_features_variant_beds.sh</code>)</p>

<h4>
<a id="collapse-site-level-features-created-above-into-gene-level-features" class="anchor" href="#collapse-site-level-features-created-above-into-gene-level-features" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Collapse site-level features created above into gene-level features</h4>

<pre><code>bash feature_construction/run_build_feature_count_summaries_all_genes.sh
</code></pre>

<p>(uses multiple cores; relies on :</p>

<ul>
<li><code>build_count_summaries_all_genes.sh</code></li>
<li><code>build_feature_summaries_all_genes.sh</code></li>
<li>
<code>build_feature_set.py</code>)</li>
</ul>

<h4>
<a id="compile-features-for-outliers-and-controls" class="anchor" href="#compile-features-for-outliers-and-controls" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Compile features for outliers and controls</h4>

<pre><code>bash feature_construction/run_compile_features_outliers.sh
</code></pre>

<p>(uses multiple cores; relies on:</p>

<ul>
<li><code>compile_features_outliers.sh</code></li>
<li><code>compile_features_outliers_nothresh.sh</code></li>
<li><code>compile_features_outliers_singletissue.sh</code></li>
<li>
<code>pick_outliers_controls_imbalanced.py</code>)</li>
</ul>

<h2>
<a id="disease-gene-annotations" class="anchor" href="#disease-gene-annotations" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Disease gene annotations</h2>

<p>We are providing the processed gene lists for the eight disease gene sets we analyzed for overlap with genes with multi-tissue outliers. <br>
We are also providing, where applicable, the commands and raw files needed to generate these processed lists.</p>

<h4>
<a id="acmg" class="anchor" href="#acmg" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>ACMG</h4>

<p>Source: <a href="http://www.ncbi.nlm.nih.gov/clinvar/docs/acmg/">http://www.ncbi.nlm.nih.gov/clinvar/docs/acmg/</a> <br>
Raw file: <code>acmg.csv</code> <br>
Download from &lt;website - coming soon&gt;. <br>
Move the downloaded file to <code>${RAREVARDIR}/features/annotations/ACMG/</code>.</p>

<h4>
<a id="clinvar" class="anchor" href="#clinvar" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>ClinVar</h4>

<p>Source: ftp://ftp.ncbi.nlm.nih.gov/pub/clinvar/gene_condition_source_id <br>
Raw file: <code>gene_condition_source_id</code> <br>
Download from &lt;website - coming soon&gt;. <br>
Move the downloaded file to <code>${RAREVARDIR}/features/annotations/ClinVar/</code>.</p>

<h4>
<a id="gwas" class="anchor" href="#gwas" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>GWAS</h4>

<p>Source: <a href="http://www.ebi.ac.uk/gwas/">http://www.ebi.ac.uk/gwas/</a> <br>
Raw file: <code>gwas_catalog_v1.0-downloaded_2015-11-30.tsv</code> <br>
Download from &lt;website - coming soon&gt;. <br>
Move the downloaded file to <code>${RAREVARDIR}/features/annotations/GWAS/</code>.</p>

<h4>
<a id="omim" class="anchor" href="#omim" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>OMIM</h4>

<p>Source: <a href="http://www.omim.org/">http://www.omim.org/</a> <br>
Raw files: <code>morbidmap.txt</code> and <code>mim2gene.txt</code> <br>
Processed file: <code>omim.genes.txt</code>
Downlaod the raw and processed files from &lt;website - coming soon&gt;. <br>
Move these files to <code>${RAREVARDIR}/features/annotations/OMIM/</code>.</p>

<p>To produce the processed file from the raw files: <br></p>

<pre><code>grep '(3)' ${RAREVARDIR}/features/annotations/OMIM/morbidmap.txt | cut -f2 | sed 's/, /\n/g' | sort | uniq &gt; ${RAREVARDIR}/features/annotations/OMIM/omim.genes.temp.txt

grep -wf ${RAREVARDIR}/features/annotations/OMIM/omim.genes.temp.txt ${RAREVARDIR}/features/annotations/OMIM/mim2gene.txt  &gt; ${RAREVARDIR}/features/annotations/OMIM/temp.mim2gene.intersection.txt

cut -f4,5 ${RAREVARDIR}/features/annotations/OMIM/temp.mim2gene.intersection.txt | sort | uniq &gt; ${RAREVARDIR}/features/annotations/OMIM/omim.genes.txt

rm ${RAREVARDIR}/features/annotations/OMIM/omim.genes.temp.txt ${RAREVARDIR}/features/annotations/OMIM/temp.mim2gene.intersection.txt
</code></pre>

<h4>
<a id="orphanet" class="anchor" href="#orphanet" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Orphanet</h4>

<p>Source: <a href="http://www.orphadata.org/data/xml/en_product6.xml">http://www.orphadata.org/data/xml/en_product6.xml</a> <br>
Raw file: <code>en_product6.xml</code> <br>
Processed file: <code>orphanet.genes.txt</code> <br>
Downlaod the raw and processed files from &lt;website - coming soon&gt;. <br>
Move these files to <code>${RAREVARDIR}/features/annotations/Orphanet/</code>.</p>

<p>To produce the processed file from the raw file: <br></p>

<pre><code>grep ENSG ${RAREVARDIR}/features/annotations/Orphanet/en_product6.xml | sort | uniq | grep -o 'ENSG[0-9]*' &gt; ${RAREVARDIR}/features/annotations/Orphanet/orphanet.genes.txt
</code></pre>

<h4>
<a id="other-cardiovascular-and-cancer-disease-genes" class="anchor" href="#other-cardiovascular-and-cancer-disease-genes" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Other: Cardiovascular and Cancer disease genes</h4>

<p>We assessed overlap of genes with multi-tissue outliers with two expert curated disease gene lists: one for heritable cancer predisposition and one for heritable cardiovascular disease. See the methods section of our manuscript for more information. <br>
Raw files: <code>cancer.genes.gold.standard.csv</code> (Cancer), <code>cardio.genes.gold.standard.csv</code> (Cardio) 
Downlaod the raw files from &lt;website - coming soon&gt;. <br>
Move these files to <code>${RAREVARDIR}/features/annotations/Other/</code>.</p>

<h2>
<a id="shared-eqtls-defined-by-metasoft" class="anchor" href="#shared-eqtls-defined-by-metasoft" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Shared eQTLs defined by METASOFT</h2>

<p>python shared.eqtls/bf.metasoft.py --META ${RAREVARDIR}/data/metasoft/Metasoft_Output_v6p.txt \
    --TISS ${RAREVARDIR}/data/metasoft/Metasoft_tissue_order.txt \
    --OUT ${RAREVARDIR}/data/metasoft/gtex.metasoft.v6p.selected.txt</p>

<h2>
<a id="main-figures" class="anchor" href="#main-figures" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Main figures</h2>

<h4>
<a id="figure-1" class="anchor" href="#figure-1" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Figure 1</h4>

<pre><code>bash paper_figures/pick.cartoon.example.sh
Rscript paper_figures/figure1a.plot.cartoon.example.R
Rscript paper_figures/figure1b.outlier.sharing.R
Rscript paper_figures/figure1c.replication.rate.consistent.R
</code></pre>

<h4>
<a id="figure-2" class="anchor" href="#figure-2" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Figure 2</h4>

<pre><code>bash paper_figures/count_rarevars.sh
Rscript paper_figures/figure2a.count.enrichments.R
Rscript paper_figures/figure2b.threshold.enrichments.R
Rscript paper_figures/figure2c.ASE.enrichments.R
Rscript paper_figures/figure2.R
</code></pre>

<h4>
<a id="figure-3" class="anchor" href="#figure-3" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Figure 3</h4>

<pre><code>Rscript paper_figures/figure3a.rare.variant.class.enrichments.R
Rscript paper_figures/figure3b.feature.enrichments.R
Rscript paper_figures/figure3de.outlier.effect.size.R
Rscript paper_figures/figure3.R
</code></pre>

<h4>
<a id="figure-4" class="anchor" href="#figure-4" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Figure 4</h4>

<pre><code>Rscript paper_figures/figure4a.uk10k.R
Rscript paper_figures/figure4b.exac.enrichments.R
Rscript paper_figures/figure4c.gene.list.enrichments.R
Rscript paper_figures/figure4.R
</code></pre>

<h4>
<a id="figure-5" class="anchor" href="#figure-5" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Figure 5</h4>

<pre><code>Rscript getPosteriorsEval.R data/genomic_features.txt data/outliers.txt
Rscript paper_figures/figure5b.R
Rscript getPosteriorsApp.R data/genomic_features.txt data/outliers.txt data/postprobs_all.txt
Rscript paper_figures/figure5c.R
</code></pre>

<h2>
<a id="supplemental-figures" class="anchor" href="#supplemental-figures" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Supplemental figures</h2>

<h4>
<a id="number-of-rare-variants-per-individual-and-pca" class="anchor" href="#number-of-rare-variants-per-individual-and-pca" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Number of rare variants per individual and PCA</h4>

<p>You need to set the path to the downloaded expression covariates and subject annotations in the script below. <br> </p>

<pre><code>Rscript paper_figures/suppfig.number.rare.vars.pca.R
</code></pre>

<h4>
<a id="improvement-of-replication-of-outliers-across-tissues-by-peer-correction" class="anchor" href="#improvement-of-replication-of-outliers-across-tissues-by-peer-correction" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Improvement of replication of outliers across tissues by PEER correction</h4>

<pre><code>Rscript paper_figures/ExtendedDataFigure3.R
</code></pre>

<h4>
<a id="distribution-of-the-nubmer-of-genes-with-an-multi-tissue-outlier" class="anchor" href="#distribution-of-the-nubmer-of-genes-with-an-multi-tissue-outlier" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Distribution of the nubmer of genes with an multi-tissue outlier</h4>

<p>You need to set the path to the subject annotations in the script below. <br> </p>

<pre><code>Rscript paper_figures/suppfig.number.outliers.per.individual.R
</code></pre>

<h4>
<a id="single-tissue-replication-analysis-controlling-for-sampling-differences" class="anchor" href="#single-tissue-replication-analysis-controlling-for-sampling-differences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Single-tissue replication analysis controlling for sampling differences</h4>

<p>Generated when running 'paper_figures/figure1b.outlier.sharing.R' above</p>

<h4>
<a id="overlap-between-single-and-multi-tissue-outliers" class="anchor" href="#overlap-between-single-and-multi-tissue-outliers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Overlap between single and multi-tissue outliers</h4>

<pre><code>Rscript paper_figures/suppfig.compare.single.multi.R
</code></pre>

<h4>
<a id="single-tissue-outlier-rare-variant-enrichments" class="anchor" href="#single-tissue-outlier-rare-variant-enrichments" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Single-tissue outlier rare variant enrichments</h4>

<p>Generated when running <code>paper_figures/figure2b.threshold.enrichments.R</code> above</p>

<h4>
<a id="multi-tissue-outlier-rare-variant-enrichments-when-excluding-coding-regions" class="anchor" href="#multi-tissue-outlier-rare-variant-enrichments-when-excluding-coding-regions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Multi-tissue outlier rare variant enrichments when excluding coding regions</h4>

<p>Generated when running <code>paper_figures/figure2a.count.enrichments.R above</code></p>

<h4>
<a id="enrichment-of-functional-genomic-annotations-among-an-expanded-set-of-multi-tissue-outliers" class="anchor" href="#enrichment-of-functional-genomic-annotations-among-an-expanded-set-of-multi-tissue-outliers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Enrichment of functional genomic annotations among an expanded set of multi-tissue outliers</h4>

<pre><code>Rscript paper_figures/ExtendedDataFigure9.R
</code></pre>

<h4>
<a id="association-between-ase-and-river-scores" class="anchor" href="#association-between-ase-and-river-scores" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Association between ASE and RIVER scores</h4>

<pre><code>Rscript paper_figures/ExtendedDataFigure11.R
</code></pre>

<h4>
<a id="correlation-between-test-posterior-probabilities-and-the-fraction-of-tissues" class="anchor" href="#correlation-between-test-posterior-probabilities-and-the-fraction-of-tissues" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Correlation between test posterior probabilities and the fraction of tissues</h4>

<pre><code>Rscript paper_figures/main_RIVER_10CV.R
Rscript paper_figures/ExtendedDataFigure12.R
</code></pre>

<h4>
<a id="distribution-of-predictive-scores-for-pathogenic-variants-and-all-variants" class="anchor" href="#distribution-of-predictive-scores-for-pathogenic-variants-and-all-variants" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Distribution of predictive scores for pathogenic variants and all variants</h4>

<pre><code>Rscript paper_figures/ExtendedDataFigure13.R
</code></pre>

<h4>
<a id="expression-levels-for-genes-proximal-to-pathogenic-variants" class="anchor" href="#expression-levels-for-genes-proximal-to-pathogenic-variants" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Expression levels for genes proximal to pathogenic variants</h4>

<pre><code>Rscript paper_figures/ExtendedDataFigure14.rpkm.R
Rscript paper_figures/ExtendedDataFigure14.Zscores.R
</code></pre>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/joed3/GTExV6PRareVariation">Gtexv6prarevariation</a> is maintained by <a href="https://github.com/joed3">joed3</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>
