---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---
## Install the dependencies
R packages
* cowplot
* data.table
* doMC
* ggplot2
* matrixStats
* peer
* plotrix
* plyr
* pROC
* RColorBrewer
* reshape2
* scales

Python modules
* numpy
* pybedtools
* pysam
* scipy

Unix packages
* GNU parallel
* tabix

External software
* bedtools v2.26.0 or later
* vcftools



## Setup
Add the following to .bashrc
```
export RAREVARDIR=<path to root of folder that contains our data>
```

Make assumed directories
```{bash sub_directories, eval=FALSE, cache=TRUE}
mkdir ${RAREVARDIR}/paper_figures
mkdir ${RAREVARDIR}/logs
mkdir ${RAREVARDIR}/preprocessingx
mkdir ${RAREVARDIR}/preprocessing/PEER
mkdir ${RAREVARDIR}/reference
mkdir ${RAREVARDIR}/data
mkdir ${RAREVARDIR}/data/GTEx
mkdir ${RAREVARDIR}/data/1KG
mkdir ${RAREVARDIR}/data/1KG_AFR
mkdir ${RAREVARDIR}/data/medz
mkdir ${RAREVARDIR}/data/singlez
# mkdir ${RAREVARDIR}/data/metasoft
# mkdir ${RAREVARDIR}/features
# mkdir ${RAREVARDIR}/features/annotations
# mkdir ${RAREVARDIR}/features/annotations/ACMG
# mkdir ${RAREVARDIR}/features/annotations/ClinVar
# mkdir ${RAREVARDIR}/features/annotations/GWAS
# mkdir ${RAREVARDIR}/features/annotations/OMIM
# mkdir ${RAREVARDIR}/features/annotations/Orphanet
# mkdir ${RAREVARDIR}/features/annotations/Other
mkdir ${RAREVARDIR}/RIVER
# mkdir ${RAREVARDIR}/RIVER/code
# mkdir ${RAREVARDIR}/RIVER/code/mfiles
# mkdir ${RAREVARDIR}/RIVER/data
# mkdir ${RAREVARDIR}/RIVER/data/expression
# mkdir ${RAREVARDIR}/RIVER/data/score
# mkdir ${RAREVARDIR}/RIVER/data/score/indiv
# mkdir ${RAREVARDIR}/RIVER/data/score/feature
# mkdir ${RAREVARDIR}/RIVER/data/rvsite
# mkdir ${RAREVARDIR}/data/wgs
```


## Download required files
Download from http://www.gtexportal.org/home/datasets: <br>
* gencode.v26.GRCh38.genes.gtf.gz
* `GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct` (gunzip it)
* `GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct` (gunzip it)
* `GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct` (gunzip it)
* `GTEx_Analysis_v8_eQTL_covariates.tar.gz` (covariates used during eQTL discovery)
* `GTEx_Analysis_v8_eQTL.tar` (eGene and significant variant-gene associations based on permutations for each tissue)
Save these files to `${RAREVARDIR}/data`
```{bash download_gtex_v8, eval=FALSE, cache=TRUE} 

# gencode.v26.GRCh38.genes.gtf
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/reference/gencode.v26.GRCh38.genes.gtf
gunzip ${RAREVARDIR}/data/GTEx/gencode.v26.GRCh38.genes.gtf.gz

# GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz
gunzip ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct.gz

# GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz
gunzip ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct.gz

# GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/rna_seq_data/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz
gunzip ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_median_tpm.gct.gz

# GTEx_Analysis_v8_eQTL_covariates.tar.gz
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL_covariates.tar.gz
tar -xzvf ${RAREVARDIR}/data/GTEx/GTEx_Analysis_v8_eQTL_covariates.tar.gz -C ${RAREVARDIR}/data/GTEx

# GTEx_Analysis_v8_eQTL.tar
wget -P ${RAREVARDIR}/data/GTEx https://storage.googleapis.com/gtex_analysis_v8/single_tissue_qtl_data/GTEx_Analysis_v8_eQTL.tar
tar -xvf ${RAREVARDIR}/data/GTEx/GTEx_Analysis_v8_eQTL.tar -C ${RAREVARDIR}/data/GTEx

# convert "c-1" files in eQTL folder into "c1"
for f in ${RAREVARDIR}/data/GTEx/GTEx_Analysis_v8_eQTL/*c-1*; do mv "$f" "$(echo "$f" | sed s/c-1/c1/)"; done

# convert "c-1" files in eQTL_covariates folder into "c1"
for f in ${RAREVARDIR}/data/GTEx/GTEx_Analysis_v8_eQTL_covariates/*c-1*; do mv "$f" "$(echo "$f" | sed s/c-1/c1/)"; done

```


Download from Gencode the v26 annotations (ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/). These gene annotations will be used to select for protein coding and lincRNA coding genes.
* `gencode.v26.annotation.gtf` (gunzip it)
```{bash gencode, eval=FALSE, cache=TRUE} 
# gencode.v26.annotation.gtf
wget -P ${RAREVARDIR}/data/GTEx ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz
zcat ${RAREVARDIR}/data/GTEx/gencode.v26.annotation.gtf.gz > ${RAREVARDIR}/data/GTEx/gencode.v26.annotation.gtf


```


Download from dbGaP: <br>
Save the following to `${RAREVARDIR}/data/GTEx`
(It is possible some of the file names may be different in the final release.)
* `GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz` (WGS for v8 individuals)
<!-- * ASE files -->
* `GTEx_Analysis_2017-06-05_v8_Annotations_SampleAttributesDS.txt` (information on tissue samples)
* `GTEx_Analysis_2017-06-05_v8_Annotations_SubjectPhenotypesDS.txt` (subject annotations, such as reported ethnicity)

Example of copying dbGaP files from locations in MARCC to `${RAREVARDIR}/data/GTEx`
```{bash dbGap, cache=TRUE, eval=FALSE}
# WGS for v8 individuals
cp /work-zfs/abattle4/bstrober/rare_variant/gtex_v8/splicing/input_data/variant_calls/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz ${RAREVARDIR}/data/GTEx/.

# information on tissue samples
cp /work-zfs/abattle4/lab_data/GTEx_v8/sample_annotations/GTEx_Analysis_2017-06-05_v8_Annotations_SampleAttributesDS.txt ${RAREVARDIR}/data/GTEx/.

# subject annotations, such as reported ethnicity
cp /work-zfs/abattle4/lab_data/GTEx_v8/sample_annotations/GTEx_Analysis_2017-06-05_v8_Annotations_SubjectPhenotypesDS.txt ${RAREVARDIR}/data/GTEx/.
```

Download from 1KG: <br>
1000 Genomes whole genome sequencing data is used to compute allele frequencies on the population of interest. Save the WGS data to `${RAREVARDIR}/data/1KG`.

`http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV/`
Integrated phased Biallelic SNP call set for the unrelated samples (N=5248)
The set is restricted to biallelic SNVs. These files contain a call set produced by aligning the 1000 Genomes Project sequence data to GRCh38 and calling directly against GRCh38. Multiple callers were used in calling, with the call sets being integrated before final genotyping and phasing.
The set is restricted to biallelic SNVs.


The following downloads the vcf files for all autosomes in release 20181203 to `${RAREVARDIR}/data/1KG`. This will take some time. 
```{bash download_1KG, eval=FALSE, cache=TRUE}
skg=${RAREVARDIR}/data/1KG

wget -P ${skg}/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV/20181203_biallelic_SNV_manifest.txt
for j in {1..22}; do
  wget -P ${skg}/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz
  wget -P ${skg}/ http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_genomes_project/release/20181203_biallelic_SNV/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz.tbi
done

# download and check manifest 
awk -v dir="$skg" '{print $3,dir"/"$1}' ${skg}/20181203_biallelic_SNV_manifest.txt > ${skg}/20181203_biallelic_SNV_manifest.md5
md5sum --check ${skg}/20181203_biallelic_SNV_manifest.md5
```

Get list of African samples from 1KG (Jessica)
```{bash download_1KG_AFR_list, eval=FALSE, cache=TRUE}
#Download sample IDs and extract African Population IDs, keeping subpop info in case it is of interest later
afr_1kg=${RAREVARDIR}/data/1KG_AFR
wget -P ${RAREVARDIR}/data/1KG/ ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/integrated_call_samples_v3.20130502.ALL.panel
awk < ${RAREVARDIR}/data/1KG/integrated_call_samples_v3.20130502.ALL.panel '($3=="AFR"){print $1, $2}' > $afr_1kg/samples_v3.20130502.AFR.txt

# get list of sample IDs only
awk '($3=="AFR"){print $1}' ${RAREVARDIR}/data/1KG/integrated_call_samples_v3.20130502.ALL.panel > $afr_1kg/afr.keep

```

<!-- Download from http://krishna.gs.washington.edu/download/CADD/v1.2/: -->
<!-- * `whole_genome_SNVs.tsv.gz` -->
<!-- * `whole_genome_SNVs.tsv.gz.tbi` -->
<!-- * `whole_genome_SNVs_inclAnno.tsv.gz` -->
<!-- * `whole_genome_SNVs_inclAnno.tsv.gz.tbi` -->

<!-- Download from ftp://ftp-trace.ncbi.nih.gov/1000genomes/ftp/release/20130502/: -->
<!-- * `ALL.chr*.phase3_shapeit2_mvncall_integrated_v5a.20130502.genotypes.vcf.gz` -->

<!-- Download from http://compbio.mit.edu/encode-motifs/: -->
<!-- * `matches.txt.gz` -->

<!-- Download Epigenomics Roadmap files from http://www.broadinstitute.org/~meuleman/reg2map/HoneyBadger2_release/DNase/p2/: -->
<!-- * `prom/BED_files_per_sample/regions_prom_E*.bed` -->
<!-- * `enh/BED_files_per_sample/regions_enh_E*.bed` -->
<!-- * `dyadic/BED_files_per_sample/regions_dyadic_E*.bed` -->

<!-- Download ExAC constraint data from ftp://ftp.broadinstitute.org/pub/ExAC_release/release0.3/functional_gene_constraint/: -->
<!-- * `forweb_cleaned_exac_r03_march16_z_data_pLI.txt` -->

<!-- Download RNA-Seq data for K562 cell lines from ENCODE at https://www.encodeproject.org/search/?type=Experiment&biosample_term_name=K562&assay_title=RNA-seq -->
<!-- * `ENCFF104VTJ_CSHL_1.tsv` -->
<!-- * `ENCFF201HGA_CSHL_2.tsv` -->
<!-- * `ENCFF387ZRA_Caltech_1.tsv` -->
<!-- * `ENCFF870QAL_Caltech_2.tsv` -->
<!-- * `ENCFF553DDU_UConn_1.tsv` -->
<!-- * `ENCFF811VBA_UConn_2.tsv` -->
<!-- * `gencode.v24.tRNAs.gtf.gz` -->

<!-- Download the hg19 human reference genome in FASTA format from http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz. Combine FASTA files for each chromosome into a single reference FASTA. -->

<!-- Download VCFs from the UK10K dataset for two studies -->
<!-- * ALSPAC VCF: https://www.google.com/url?q=https://www.ebi.ac.uk/ega/datasets/EGAD00001000740&sa=D&ust=1507684754410000&usg=AFQjCNFJ8j8ESpELYWdyx6wU7jGU1w2waQ -->
<!-- * TWINS UK VCF: https://www.google.com/url?q=https://www.ebi.ac.uk/ega/datasets/EGAD00001000741&sa=D&ust=1507684754411000&usg=AFQjCNG16PtulHOxP-G1QN5knTBs_wlFTQ -->

<!-- Other files: <br> -->
<!-- * `gtex.lumpy.gs.svscore.low_conf.vcf.gz` (Obtained directly from the Hall Lab.) -->

<!-- Download annotation resources to generate genomic features for running RIVER -->
<!-- * CADD: http://krishna.gs.washington.edu/download/CADD/v1.3/whole_genome_SNVs_inclAnno.tsv.gz and http://krishna.gs.washington.edu/download/CADD/v1.3/whole_genome_SNVs_inclAnno.tsv.gz.tbi -->
<!-- * DANN: https://cbcl.ics.uci.edu/public_data/DANN/data/DANN_whole_genome_SNVs.tsv.bgz and https://cbcl.ics.uci.edu/public_data/DANN/data/DANN_whole_genome_SNVs.tsv.bgz.tbi -->
<!-- * phylop: http://hgdownload.cse.ucsc.edu/goldenpath/hg19/phyloP100way/hg19.100way.phyloP100way.bw -->
<!-- * chromHMM: http://hgdownload.soe.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeBroadHmm/wgEncodeBroadHmmGm12878HMM.bed.gz -->


# Pipeline

## GTEx v8
### Expression data correction and normalization

#### Generate rpkm and read count matrices from GTEx V8 combined file
```{bash tpkm_read_count, eval=FALSE, cache=TRUE}
cat ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_Annotations_SampleAttributesDS.txt | \
	cut -f1,14 | sed 's/ - /_/' | sed 's/ /_/g' | sed 's/(//' | sed 's/)//' | sed 's/c-1/c1/' | grep -v NA12878 > \
	${RAREVARDIR}/reference/gtex_v8_samples_tissues.txt

python preprocessing/split_by_tissues.py \
    --GTEX ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_tpm.gct \
    --SAMPLE ${RAREVARDIR}/reference/gtex_v8_samples_tissues.txt \
    --OUT ${RAREVARDIR}/preprocessing/PEER \
    --END .tpm.txt

python preprocessing/split_by_tissues.py \
    --GTEX ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_RNASeQCv1.1.9_gene_reads.gct \
    --SAMPLE ${RAREVARDIR}/reference/gtex_v8_samples_tissues.txt \
    --OUT ${RAREVARDIR}/preprocessing/PEER \
    --END .reads.txt
```

#### Run bash scripts to generate PEER corrected data (includes non-EAs) with covariates removed
(Uses multiple cores. Currently set to 10 cores. Can be altered by changing the number of cores <br> 
specified by `parallel --jobs 10` in the scripts `preprocessing/PEER/calc.PEER.factors.all.tissues.sh` and <br>
`preprocessing/PEER/calc.residuals.sh`)
```{bash peer_correction, eval=FALSE, cache=TRUE}
# Make list of tissues
ls ${RAREVARDIR}/data/GTEx/GTEx_Analysis_v8_eQTL_covariates | cut -d "." -f1 > ${RAREVARDIR}/reference/gtex_v8_tissues_all_normalized_samples.txt

# Run PEER pipeline
bash preprocessing/PEER/PEER.pipeline.sh
```

<!-- Obtain gene expression data that is corrected and normalized from Nicole. This data was saved to `$RAREVARDIR/preprocessing/PEER_nicole` -->
<!-- ```{bash nicole_data, eval=FALSE, cache=TRUE} -->
<!-- mkdir ${RAREVARDIR}/preprocessing/PEER_nicole -->
<!-- cp /work-zfs/abattle4/bstrober/rare_variant/gtex_v8/splicing/unsupervised_modeling/unsupervised_learning_input/total_expression_data_from_nicole/* ${RAREVARDIR}/preprocessing/PEER_nicole/. -->
<!-- ``` -->


<!-- Convert the expression data table from long to wide. -->
<!-- ```{bash long_to_wide, eval=FALSE, cache=TRUE} -->
<!-- infolder=$RAREVARDIR/preprocessing/PEER_nicole -->
<!-- outfolder=$RAREVARDIR/preprocessing/PEER -->
<!-- for file in $(ls ${infolder}/*globalOutliers.v8ciseqtls.removed.outlierGenes.zscores.txt); do -->
<!-- base=$(basename $file) -->
<!-- echo ${file} -->
<!-- Rscript --vanilla preprocessing/long_to_wide_exprbytissue.R ${file} ${outfolder}/${base} -->
<!-- done -->
<!-- ``` -->

### Make list of individuals
Make list of all individuals from GTEx v8. Make list of all individuals with reported African American ancestry and European ancestry.
* `${RAREVARDIR}/reference/gtex_v8_individuals_all_normalized_samples.txt` - All 948 individuals from GTEx v8
* `${RAREVARDIR}/reference/gtex_v8_wgs_individuals.txt` - All 838 individuals from GTEx v8 with WGS data
* `${RAREVARDIR}/reference/gtex_v8_individuals_AFA.txt` - 121 African American individuals from GTEx v8
* `${RAREVARDIR}/reference/gtex_v8_wgs_individuals_AFA.txt` - 103 African American individuals from GTEx v8 with WGS data
* `${RAREVARDIR}/reference/gtex_v8_individuals_EUR.txt` - 804 European individuals from GTEx v8
* `${RAREVARDIR}/reference/gtex_v8_wgs_individuals_EUR.txt` - 714 European individuals from GTEx v8 with WGS 
```{bash gtex_individuals, eval=FALSE, cache=TRUE}
# All individuals
## list of all individuals from GTEx v8 (some do not have WGS data)
cat ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_Annotations_SubjectPhenotypesDS.txt | awk -F "\t" 'NR>1{print $1}' \
> ${RAREVARDIR}/reference/gtex_v8_individuals_all_normalized_samples.txt

## list of all individuals from GTEx v8 with WGS data
vcf-query -l ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_WholeGenomeSeq_838Indiv_Analysis_Freeze.SHAPEIT2_phased.vcf.gz \
> ${RAREVARDIR}/reference/gtex_v8_wgs_individuals.txt

# African American individuals
## list of all African American individuals from GTEx v8 (some do not have WGS data)
cat ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_Annotations_SubjectPhenotypesDS.txt | awk -F '\t' '{if ($5 == 2) print $1;}' \
> ${RAREVARDIR}/reference/gtex_v8_individuals_AFA.txt

## list of all African American individuals from GTEx v8 with WGS data
awk 'NR==FNR { lines[$0]=1; next } $0 in lines' ${RAREVARDIR}/reference/gtex_v8_wgs_individuals.txt ${RAREVARDIR}/reference/gtex_v8_individuals_AFA.txt > ${RAREVARDIR}/reference/gtex_v8_wgs_individuals_AFA.txt


# European individuals
# list of all European individuals from GTEx v8 (some do not have WGS data)
cat ${RAREVARDIR}/data/GTEx/GTEx_Analysis_2017-06-05_v8_Annotations_SubjectPhenotypesDS.txt | awk -F '\t' '{if ($5 == 3) print $1;}' \
> ${RAREVARDIR}/reference/gtex_v8_individuals_EUR.txt

# list of all African American individuals from GTEx v8 with WGS data
awk 'NR==FNR { lines[$0]=1; next } $0 in lines' ${RAREVARDIR}/reference/gtex_v8_wgs_individuals.txt ${RAREVARDIR}/reference/gtex_v8_individuals_EUR.txt > ${RAREVARDIR}/reference/gtex_v8_wgs_individuals_EUR.txt
```

<!-- ### Make list of tissues (based on Nicole's data) -->
<!-- ```{bash , eval=FALSE, cache=TRUE} -->
<!-- ls ${RAREVARDIR}/preprocessing/PEER | cut -d. -f1 \ -->
<!-- > ${RAREVARDIR}/reference/gtex_v8_tissues_all_normalized_samples.txt -->
<!-- ``` -->

### Make flat files from normalized data
```{bash, eval=FALSE, cache=TRUE}
python preprocessing/gather_filter_normalized_expression_v8.py
```

### Preparing gene annotations from gencode
This list will be used to filter out genes for those that are protein coding and lincRNA coding
**Resulting `reference/gencode.v26.genes.v8.patched_contigs_genetypes_autosomal.txt` may be empty for some reason**
```{bash gencode_genes, eval=FALSE, cache=TRUE}
bash preprocessing/process.reference.files_v8.sh ${RAREVARDIR}/data/GTEx/gencode.v26.GRCh38.genes.gtf.gz ${RAREVARDIR}/download/gencode/gencode.v26.annotation.gtf
```

### Call multi-tissue outliers
For each gene, the individual with the most extreme median Z-score taken across tissues was identified as a multi-tissue outlier for that gene provided the absolute median Z-score was at least 2
```{bash multi_outliers, eval=FALSE, cache=TRUE}
Rscript call_outliers/call_outliers_medz_v8.R
```
Multi-tissue outlier calling produces the following files
#### data/outliers_medz_counts.txt
Description: Matrix of genes by individuals representing the number of tissues used in computing the median Z-score for each (individual, gene) pair
File format: Tab delimited file where rows are genes (ENSG IDs) and columns are individuals (GTEx individual IDs). Values represent the number of tissues for the given (individual, gene) pair used in computation of the median Z-score

#### data/outliers_medz_zscores.txt
Description: Matrix of genes by individuals holding the median Z-score for each (individual, gene) pair
File format: Tab delimited file where rows are genes (ENSG IDs) and columns are individuals (GTEx individual IDs). Values represent the median Z-score for the given (individual, gene) pair. For (individual, gene) pairs where the number of available tissues is less than 5, NA is used to indicate the outlier test was not performed

#### data/outliers_medz_picked.txt
Description: Most extreme multi-tissue outliers for each gene with criterion |median Z-score| > 2
File format: Tab delimited file where first column is the gene ENSG ID, second column GTEx individual ID, third column the number of tissues included in the median Z-score computation, and the fourth column the median Z-score

#### data/outliers_medz_picked_counts_per_ind.txt
Description: Number of genes where the GTEx individual is a multi-tissue outlier
File format: Tab delimited file where the first column is the GTEx individual ID and the second column is the number of genes where that individual is a multi-tissue outlier

#### preprocessing/gtex_v8_wgs_ids_outlier_filtered.txt
Description: list of GTEx v8 individuals that have WGS data and are multi-tissue outliers for at least one gene.

<!-- ### Call single-tissue outliers -->
<!-- For each tissue, an individual was called a single-tissue outlier for a particular gene if that individual had the largest absolute Z-score and the absolute value was at least 2 -->
<!-- ```{bash single_outliers, eval=FALSE, cache=TRUE} -->
<!-- python call_outliers/call_outliers_single_tissue_v8.py -->
<!-- ``` -->


## 1KG

### Subset 1KG
Subset 1KG for African samples (look into doing this with bcfTools)
```{bash subset_1KG, eval=FALSE, cache=TRUE}
skg=${RAREVARDIR}/data/1KG
afr_1kg=${RAREVARDIR}/data/1KG_AFR

bgzip -fdc ${skg}/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf.gz > ${skg}/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf
sed 's/^##fileformat=VCFv4.3/##fileformat=VCFv4.2/' ${skg}/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129.phased.vcf  > ${skg}/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129_v42.phased.vcf

vcftools --vcf ${skg}/ALL.chr${j}.shapeit2_integrated_v1a.GRCh38.20181129_v42.phased.vcf \
  --keep ${afr_1kg}/afr.keep \
  --remove-indels --remove-filtered-all \
  --max-missing-count 10 \
  --not-chr X --recode \
  --recode-INFO-all \
  --stdout >  ${afr_1kg}/AFR_1KG_chr${j}.vcf
  
bgzip -c ${afr_1kg}/AFR_1KG_chr${j}.vcf > ${afr_1kg}/AFR_1KG_chr${j}.vcf.gz
tabix -f ${afr_1kg}/AFR_1KG_chr${j}.vcf.gz
```


### Get frequency 

## Use GTEx and 1KG to get rare variants

## Analysis of African American population
Get number of rare variants (not neccessarily those within 10kb of target genes)
see
RIVER/rare_variants_per_indiv_jb.sh

# Making Figures

## Reproduced figures from mini project 2
1a on v6 data showing median z scores across tissues
```{bash}
cd paper_figures
bash pick.cartoon.example.sh
Rscript figure1a.plot.cartoon.example.R
```

1c on v6 data showing replication ratesâ€¦ very different values
```{bash}
Rscript call_outliers/multi_tissue_replication.R
Rscript paper_figures/figure1c.replication.rate.consistent.R
```

## Reproduced Figures v6 --> v8
1a on v8 data
```{bash}
cd paper_figures
bash pick.cartoon.example_v8.sh
Rscript figure1a.plot.cartoon.example_v8.R
```

1c on v8 data
```{bash}
Rscript call_outliers/multi_tissue_replication_v8.R
Rscript paper_figures/figure1c.replication.rate.consistent_v8.R
```

## Figures from mini project 3
PCA highlighting AFR population 

Differences between V6 and V8 Multitissue outliers
```{bash}
python call_outliers/outliers_histogram.py
```

## RIVER analysis for AFR population
Allele Freq box plot demonstrating how many more rare alleles there are in AFA population (one box for european and one box for african)
```{bash}
Rscript RIVER/compare_rare_variants.R
```

Overlap of rare variant counts

AUC (RIVER model evaluation)

Table of features used in EUR-ALL/AFR-ALL/COMMON

RIVER eval on European, all features

RIVER eval on European, common features

RIVER eval on African, all features

RIVER eval on African, common features


## Comparison between AFR and EUR in v8
Modified version of Fig. 5a from watershed paper 
Outliers section (# of outliers per individual) 
African and european
Watershed > 0.5 section (# of watershed scores >0.5 per individual)
African and european
Watershed > 0.9 section (# of watershed scores >0.9 per individual)
African and european

Subset multi-tissue outliers count per individual by European and African populations.
* `${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind_EUR.txt` - Outlier count per European individual
* `${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind_AFA.txt` - Outlier count per African individual
```{bash}
# European Outliers
grep -f ${RAREVARDIR}/reference/gtex_v8_individuals_EUR.txt ${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind.txt > ${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind_EUR.txt

# African American outliers
grep -f ${RAREVARDIR}/reference/gtex_v8_individuals_AFA.txt ${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind.txt > ${RAREVARDIR}/archive/data/v8/outliers_medz_picked_counts_per_ind_AFA.txt
```
