# =======================================================================================================
# This is a script for generating EDF10b-d: Distribution of predictive scores for pathogenic variants 
#         and all variants
#
# input:
#       1. figure5c.out.RData
#          Data generated by running figure5c.R
#
# output:
#       1. EDF10bd.i.pdf (i: b-d separately)
#       2. EDF10bd.out.RData
#          Store all data used and generated by running this script
#
# Note that the final figure was slightly modified by using inkscape for visualization purposes
#
# =======================================================================================================

#!/usr/bin/env Rscript

# Recall required packages
rm(list=ls(all=TRUE))

# Master directory
dir = Sys.getenv('RAREVARDIR')

# Recall required functions
source("getPlots.R") # this code should be in the same directory as the directory for this script



# =========== Load input data
load(file = paste(dir,"/data/figure5c.out.RData",sep=""))



# =========== Main
##### Distribution of |median Z-score| for all instances and instances with 27 pathogenic variants
par(mar=c(6.1, 6.1, 4.1, 4.1))
dat = rbind(data.frame(x=log10(postprobs[,"AbsMed"]), group="A"), # background
            data.frame(x=log10(clinvar.pathogenic[,"AbsMed"]), group="B")) # pathogenic

getHistExmpls2(dat, "Gene expression, log10(|Median Z-score|)", min(dat[,"x"]), max(dat[,"x"]))

# P-value of comparing |median Z-score| between background and ClinVar-matched samples
wilcox.test(postprobs[,"AbsMed"],clinvar.pathogenic[,"AbsMed"])$p.value 

ggsave(paste(dir,"/paper_figures/EDF10bd.b.pdf",sep=""),width=6,height=3)

##### Distribution of P(FR | G) for all instances and instances with 27 pathogenic variants
par(mar=c(6.1, 6.1, 4.1, 4.1))
dat = rbind(data.frame(x=log10(postprobs[,"pZgivenG"]), group="A"), # background
            data.frame(x=log10(clinvar.pathogenic[,"pZgivenG"]), group="B")) # pathogenic

getHistExmpls2(dat, "Genomic annotations, log10(P(FR | G))", min(dat[,"x"]), max(dat[,"x"]))

# P-value of comparing P(FR | G) between background and ClinVar-matched samples
wilcox.test(postprobs[,"pZgivenG"],clinvar.pathogenic[,"pZgivenG"])$p.value 

ggsave(paste(dir,"/paper_figures/EDF10bd.c.pdf",sep=""),width=6,height=3)

##### Distribution of P(FR | G, E) for all instances and instances with 27 pathogenic variants
par(mar=c(6.1, 6.1, 4.1, 4.1))
dat = rbind(data.frame(x=log10(postprobs[,"pZgivenGE"]), group="A"), # background
            data.frame(x=log10(clinvar.pathogenic[,"pZgivenGE"]), group="B")) # pathogenic

getHistExmpls2(dat, "RIVER, log10(P(FR | G, E))", min(dat[,"x"]), max(dat[,"x"]))

# P-value of comparing P(FR | G, E) between background and ClinVar-matched samples
wilcox.test(postprobs[,"pZgivenGE"],clinvar.pathogenic[,"pZgivenGE"])$p.value 

ggsave(paste(dir,"/paper_figures/EDF10bd.d.pdf",sep=""),width=6,height=3)



# =========== Save data
save.image(file = paste(dir,"/data/EDF10bd.out.RData",sep=""))